base/compiler/ssair/passes.jl

```diff
@@ -97,20 +97,39 @@ end
 # if this load at `idx` have any "safe" `setfield!` calls that define the field
 function has_safe_def(
     ir::IRCode, domtree::DomTree, allblocks::Vector{Int}, du::SSADefUse,
-    newidx::Int, idx::Int, inclusive::Bool = false)
-    def = first(find_def_for_use(ir, domtree, allblocks, du, idx, inclusive))
-
-    # this field is supposed to be defined at the `:new` site (but it's not and thus this load will throw)
+    newidx::Int, idx::Int)
+    def, _, _ = find_def_for_use(ir, domtree, allblocks, du, idx)
+    # will throw since we already checked this `:new` site doesn't define this field
     def == newidx && return false
+    # found a "safe" definition
+    def ≠ 0 && return true
+    # we may still be able to replace this load with `PhiNode`
+    block = block_for_inst(ir, idx)
+    seen = BitSet(block)
+    return has_safe_def(ir, domtree, allblocks, du, newidx, block, seen)
+end
 
-    def ≠ 0 && return true # found a "safe" definition
-
-    # we may be able to replace this load with `PhiNode` if all the predecessors have "safe" definitions
-    idxblock = block_for_inst(ir, idx)
-    for pred in ir.cfg.blocks[idxblock].preds
-        lastidx = last(ir.cfg.blocks[pred].stmts)
-        # NOTE `lastidx` isn't a load, thus we can use inclusive coondition within the `find_def_for_use`
-        has_safe_def(ir, domtree, allblocks, du, newidx, lastidx, true) || return false
+# recursively examine if all predecessors of `block` have any "safe" definition
+# XXX this implementation avoids infinite cycles by maintaining a sigle set of seen basic blocks,
+# and obviously it doesn't handle cycles very well
+function has_safe_def(
+    ir::IRCode, domtree::DomTree, allblocks::Vector{Int}, du::SSADefUse,
+    newidx::Int, block::Int, seen::BitSet)
+    preds = ir.cfg.blocks[block].preds
+    isempty(preds) && return false
+    for pred in preds
+        # if this block has already been examined, we should bail out to avoid infinite cycles
+        pred in seen && return false
+        idx = last(ir.cfg.blocks[pred].stmts)
+        # NOTE `idx` isn't a load, thus we can use inclusive coondition within the `find_def_for_use`
+        def, _, _ = find_def_for_use(ir, domtree, allblocks, du, idx, true)
+        # will throw since we already checked this `:new` site doesn't define this field
+        def == newidx && return false
+        # found a "safe" definition for this predecessor
+        def ≠ 0 && continue
+        # check for the predecessors of this predecessor
+        push!(seen, pred)
+        has_safe_def(ir, domtree, allblocks, du, newidx, pred, seen) || return false
```

Comment:
Side Note: In a compiler it's generally best to avoid recursion and use worklists instead. 

Recursion can always blow-up the stack on a sufficiently complex input program

Reply:
Good to know, I don't read much compiler books yet, but now I see why passes.jl is full of worklists :)
Well, does this imply that in Julia at least, it's generally better to switch to iteration if we don't spend much attention to readability or such, but only care about performance?